### DISCUSSION TITLE
Limit results by unique groups (multiple records per group)

### DISCUSSION THREAD

**Discussion by mdonkers** [ClickHouse/ClickHouse#65652] [Q&A] [answered]
Hi folks,

I'm having a hard time to figure out an efficient query I believe should be possible.
Perhaps someone has better ideas.

We're doing a `GROUP BY` with each 'group' returning 0 or more records. And would like to limit the number of groups returned. `LIMIT x BY` doesn't work, because it would limit the records per group, not the number of groups itself.

Currently we only thought of using a `WINDOW` function, but it still requires ClickHouse to process all data in range:

```sql
SELECT
  Name,
  COUNT(),
  (intDiv(Timestamp, 60) * 60) as Bucket,
  dense_rank() OVER(ORDER BY Name) AS Rank
FROM %[1]s.%[2]s
WHERE Rank <= 1000
GROUP BY Name, Bucket, Rank
ORDER BY Name, Bucket
```

In the above case, we want to have the count of records per name, per 1 minute. But we only need the first 1000 unique names, no need to do further processing. Also it doesn't need to be the 1000 'Name's with the most records. But if we get values for a 'Name', we do need all 'buckets' within the given time-frame.

Any ideas how else this might be possible?

(data is ordered by `Name` in the table. So depending on the limit, I believe it might save quite some resources for us if ClickHouse could stop early)

**Comment by UnamedRus** [comment, marked as answer]
│           │           0 │
│ max_rows_to_group_by                       │ 0        │       0 │ If aggregation during GROUP BY is generating more than the specified number of rows (unique GROUP BY keys), the behavior will be determined by the 'group_by_overflow_mode' which by default is - throw an exception, but can be also switched to an approximate GROUP BY mode. │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ UInt64              │ 0        │           │           0 │
│ group_by_overflow_mode                     │ throw    │       0 │ What to do when the limit is exceeded.                                                                                                                                                                                                                     │ ᴺᵁᴸᴸ │ ᴺᵁᴸᴸ │        0 │ OverflowModeGroupBy │ throw    │           │           0 │

group_by_overflow_mode = 'any'  

OR 


group_by_overflow_mode = 'break' with optimize_aggregation_in_order=1  

Or make subquery, to get list of 1000 names and use it as FILTER IN statement for your main GROUP BY query

**Reply by mdonkers** [reply]
Thanks!
Unfortunately the `max_rows_to_group_by` doesn't seem to be as accurate as we need.
But then a subquery makes more sense and works for our case.
